### DISCUSSION TITLE
My planned PRs

### DISCUSSION THREAD

**Discussion by odenix** [apple/pkl#834] [General]
Here is a list of things I've been working on. I'm hesitant to submit more PRs before my pending PRs have been reviewed, but I thought that announcing my plans might help with coordinating efforts.

1. Generate unified API docs (Java + Kotlin) with [Dokka](https://github.com/Kotlin/dokka)
  Status: finished
1. Update Kotlin to 2.0
  Status: finished
1. Replace `VmObjectLike.iterateMemberValues(Consumer)` with external iterators for properties, elements, entries, and members
   This solves many current issues, both user-facing and internal. Example:
   ```
   for (var cursor = object.properties(CursorOption.ANY_ORDER); cursor.advance(); ) {
     var key = cursor.key(); var value = cursor.value(); ...
   }  
   ```
   Status: finished
1. codegen-java: Generate record classes
  Status: in progress (most effort is in testing)
1. codegen-java: Support `@Nullable` annotations
  Status: in progress, blocked on feedback on https://github.com/apple/pkl/issues/811
1. Improve implementation of `TypeNode`
  In particular, `TypeNode.executeEagerly` should be replaced with `EagerListingTypeNode` and `EagerMappingTypeNode`.
  This will also enable to bring back specializations that had to be removed when `executeEagerly` was introduced.
  Status: started
1. Switch to `org.jspecify.annotations.Nullable`
  Status: not started, best done after (2) has landed
1. Replace `EconomicMap.create()` with `EconomicMap.emptyMap()` for maps that are never mutated
  Status: abandoned, didn't seem worth the PR overhead

**Comment by StefMa** [comment]
In regards of dokka, did u saw this https://kotlinlang.org/docs/dokka-migration.html? Dokka (1.x) seems to be a bit outdated and they changed quite a lot in version 2.
Maybe it make sense to postpone this a bit until 2.0 is more production ready? 🤔 🤷‍♂️

**Reply by odenix** [reply]
My PR uses [dokkatoo](https://github.com/adamko-dev/dokkatoo), which is the production-ready foundation of the upcoming Dokka 2.0 Gradle plugin. It works great, and I don't think there's a need to wait.

**Reply by StefMa** [reply]
Nice 👍then you're set 🙃 

Just wanted to point this out in case you missed it. Would be sad if we had to rework this in the (near) future.

**Comment by bioball** [comment]
Thanks for letting us know! Also, somehow I missed this topic while I was out, apologies for the late response.

These PRs all look good! I don't think you need to wait to submit these PRs. My immediate goal is to review your existing PRs, and I don't know if having these future PRs in really changes anything. If anything, it helps us understand where you would like to take things (especially a PR that replaces `VmObjectLike.iterateMemberValues`).

**Comment by bioball** [comment]
> Replace VmObjectLike.iterateMemberValues(Consumer) with external iterators for properties, elements, entries, and members
This solves many current issues, both user-facing and internal.

@odenix : do you mind submitting the PR for this?

**Reply by odenix** [reply]
@bioball Here is the PR I had prepared (in Nov 2024): https://github.com/odenix/pkl/commit/e469f5424d41f24e1a3fb8cbef6e2ba395b9df99

Can you take over from here? Happy to answer questions.

**Reply by bioball** [reply]
Yes, thanks!
